libraries/integration/Microsoft.Bot.Builder.ApplicationInsights.Core/TelemetrySaveBodyASPMiddleware.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Memory;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.ApplicationInsights.Core
+{
+    public class TelemetrySaveBodyASPMiddleware
+    {
+        private readonly RequestDelegate _next;
+
+        public TelemetrySaveBodyASPMiddleware(RequestDelegate next)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (context.Request.Method == "POST")
+            {
+                var memoryCache = context.RequestServices.GetService(typeof(IMemoryCache)) as IMemoryCache;
+
+                if (memoryCache != null)
+                {
+                    context.Request.EnableBuffering();
+                    try
+                    {
+                        using (var reader = new StreamReader(context.Request.Body, Encoding.UTF8, true, 1024, true))
+                        {
+                            // Set cache options.
+                            var cacheEntryOptions = new MemoryCacheEntryOptions()
+                                .SetSlidingExpiration(TimeSpan.FromSeconds(3)); // Keep in cache for this time, reset time if accessed.
+
+                            var body = reader.ReadToEnd();
+                            var jsonObject = JObject.Parse(body);
```

Comment:
⚠️ If you were to keep this implementation (which I don't suggest you do)... ⚠️ 

Ok, so... I don't see the value in parsing the JSON here at all yet. You only need to understand the JSON in the `TelemetryBotIdInitializer` itself. I would just read the body as a string and defer JSON parsing to that class:

```CSharp
public async Task Invoke(HttpContext context)
{
    var request = context.Request;

    // Only inspect the request if it's a POST of application/json content
    if (request.Method == "POST"
            &&
        request.GetTypedHeaders().ContentType.MediaType == "application/json")
    {
        // Enable buffering so downstream consumers can also read the body
        request.EnableBuffering();

        var body = request.Body;

        try
        {
            using (var reader = new StreamReader(body, Encoding.UTF8, detectEncodingFromByteOrderMarks: true, bufferSize: 1024, leaveOpen: true))
            {
                context.Items[TelemetryBotIdInitializer.HttpContextItemsKey] = await reader.ReadToEndAsync();
            }
        }
        finally
        {
            // Rewind body stream for downstream consumers
            body.Position = 0;
        }
    }

    await _next(context);
}
```
Then, as I said, I would worry about pulling the string back out and turning it into a `JObject` in `TelemetryBotIdInitializer`.